### DISCUSSION TITLE
Allocation in threaded or multi-core code

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#12561] [General]
Are the following statements true?
 1. Code running on another core must not allocate.
 2. On a single core, code running in a spawned thread must not allocate.

**Comment by dlech** [comment]
I don't think so. If GIL is enabled, then allocation is protected by the GIL. If GIL is disabled, then `GC_ENTER()/GC_EXIT()` are defined as a global `gc_mutex` lock/unlock to ensure synchronized access to the GC.

**Reply by jimmo** [reply]
Yep that's right. In both cases, GIL (esp32) or no-GIL (unix/rp2), any Python code can allocate regardless of which thread.
